utxo/src/utxo_impl/utxo_storage.rs

```diff
@@ -0,0 +1,531 @@
+}
+
+pub struct UtxoDB<'a, S: UtxosPersistentStorage>(&'a mut S);
+
+impl<'a, S: UtxosPersistentStorage> UtxoDB<'a, S> {
+    pub fn new(store: &'a mut S) -> Self {
+        Self(store)
+    }
+
+    pub fn set_undo_data(&mut self, id: Id<Block>, undo: &BlockUndo) -> Result<(), crate::Error> {
+        self.0.set_undo_data(id, undo)
+    }
+
+    pub fn del_undo_data(&mut self, id: Id<Block>) -> Result<(), crate::Error> {
+        self.0.del_undo_data(id)
+    }
+
+    pub fn get_undo_data(&self, id: Id<Block>) -> Result<Option<BlockUndo>, crate::Error> {
+        self.0.get_undo_data(id)
+    }
+}
+
+impl<'a, S: UtxosPersistentStorage> UtxosView for UtxoDB<'a, S> {
+    fn get_utxo(&self, outpoint: &OutPoint) -> Option<Utxo> {
+        match self.0.get_utxo(outpoint) {
+            Ok(res) => res,
+            Err(e) => {
+                panic!(
+                    "Database error while attempting to retrieve utxo from the database: {}",
+                    e
+                );
+            }
+        }
+    }
+
+    fn has_utxo(&self, outpoint: &OutPoint) -> bool {
+        self.get_utxo(outpoint).is_some()
+    }
+
+    fn get_best_block_hash(&self) -> Option<Id<Block>> {
+        match self.0.get_best_block_id() {
+            Ok(opt_id) => opt_id,
+            Err(e) => {
+                panic!(
+                    "Database error while attempting to retrieve utxo set best block hash from the database: {}",
+                    e
+                );
+            }
+        }
+    }
+
+    fn estimated_size(&self) -> Option<usize> {
+        None
+    }
+
+    fn derive_cache(&self) -> UtxosCache {
+        let mut cache = UtxosCache::new(self);
+        if let Some(hash) = self.get_best_block_hash() {
+            cache.set_best_block(hash);
+        }
+        cache
+    }
+}
+
+impl<'a, S: UtxosPersistentStorage> FlushableUtxoView for UtxoDB<'a, S> {
+    fn batch_write(
+        &mut self,
+        utxos: crate::utxo_impl::ConsumedUtxoCache,
+    ) -> Result<(), crate::Error> {
+        // check each entry if it's dirty. Only then will the db be updated.
+        for (key, entry) in utxos.container {
+            let outpoint = &key;
+            if entry.is_dirty() {
+                if let Some(utxo) = entry.utxo() {
+                    self.0.set_utxo(outpoint, utxo)?;
+                } else {
+                    // entry is spent
+                    self.0.del_utxo(outpoint)?;
+                };
+            }
+        }
+        self.0.set_best_block_id(&utxos.best_block)?;
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+struct UtxoInMemoryDBImpl {
+    store: BTreeMap<OutPoint, Utxo>,
+    undo_store: HashMap<H256, BlockUndo>,
+    best_block_id: Option<Id<Block>>,
+}
+
+impl UtxoInMemoryDBImpl {
+    fn new() -> Self {
+        Self {
+            store: BTreeMap::new(),
+            undo_store: HashMap::new(),
+            best_block_id: None,
+        }
+    }
+}
+
+impl UtxosPersistentStorage for UtxoInMemoryDBImpl {
```

Comment:
Unless I missed something, none of these functions ever return an error so it doesn't really make sense to return a `Result<>`, unless it's of course required by the database interface.